File: src/jp_quality_gate/text.py
from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Iterator

_INLINE_CODE_RE = re.compile(r"(`+)([^\n]*?)\1")
_URL_RE = re.compile(r"https?://[^\s<>()]+")
_SENTENCE_BOUNDARY_RE = re.compile(r"[\n。！？!?]+")
_CLAUSE_BOUNDARY_RE = re.compile(r"[、，,:：；;（）()［］\[\]{}]+")


@dataclass(frozen=True, slots=True)
class Segment:
    start: int
    end: int
    text: str


def _blank_preserve_newlines(text: str) -> str:
    return "".join("\n" if char == "\n" else " " for char in text)


def mask_markdown_nonprose(text: str, *, include_code: bool = False) -> str:
    """Mask Markdown code and URLs while preserving offsets and newlines."""
    if include_code:
        masked = text
    else:
        lines = text.splitlines(keepends=True)
        output: list[str] = []
        fence_char: str | None = None
        fence_len = 0

        for line in lines:
            stripped = line.lstrip()
            match = re.match(r"(`{3,}|~{3,})", stripped)

            if fence_char is None and match:
                token = match.group(1)
                fence_char, fence_len = token[0], len(token)
                output.append(_blank_preserve_newlines(line))
                continue

            if fence_char is not None:
                output.append(_blank_preserve_newlines(line))
                closing = re.match(rf"{re.escape(fence_char)}{{{fence_len},}}", stripped)
                if closing:
                    fence_char = None
                    fence_len = 0
                continue

            output.append(line)

        masked = "".join(output)
        masked = _INLINE_CODE_RE.sub(lambda m: _blank_preserve_newlines(m.group(0)), masked)

    masked = _URL_RE.sub(lambda m: " " * len(m.group(0)), masked)
    return masked


def is_cjk_relevant(char: str) -> bool:
    cp = ord(char)
    return (
        0x3040 <= cp <= 0x30FF  # Hiragana + Katakana
        or 0x3400 <= cp <= 0x4DBF  # CJK Extension A
        or 0x4E00 <= cp <= 0x9FFF  # CJK Unified Ideographs
        or 0xF900 <= cp <= 0xFAFF  # CJK Compatibility Ideographs
        or 0x20000 <= cp <= 0x323AF  # CJK Extensions B-I (broad range)
    )


def count_cjk_relevant(text: str) -> int:
    return sum(1 for char in text if is_cjk_relevant(char))


def _trim_span(text: str, start: int, end: int) -> tuple[int, int] | None:
    while start < end and text[start].isspace():
        start += 1
    while end > start and text[end - 1].isspace():
        end -= 1
    return (start, end) if start < end else None


def _split_spans(text: str, start: int, end: int, boundary: re.Pattern[str]) -> Iterator[tuple[int, int]]:
    cursor = start
    for match in boundary.finditer(text, start, end):
        trimmed = _trim_span(text, cursor, match.start())
        if trimmed:
            yield trimmed
        cursor = match.end()
    trimmed = _trim_span(text, cursor, end)
    if trimmed:
        yield trimmed


def iter_cj_segments(masked_text: str, *, min_cjk: int = 4) -> list[Segment]:
    """Return sentence and clause spans suitable for CJClassifier.

    Smaller clause spans are included to catch a Chinese phrase embedded in an
    otherwise Japanese sentence. Duplicates are removed while offsets are kept.
    """
    spans: set[tuple[int, int]] = set()

    for sentence_start, sentence_end in _split_spans(
        masked_text, 0, len(masked_text), _SENTENCE_BOUNDARY_RE
    ):
        sentence = masked_text[sentence_start:sentence_end]
        if count_cjk_relevant(sentence) >= min_cjk:
            spans.add((sentence_start, sentence_end))

        for clause_start, clause_end in _split_spans(
            masked_text, sentence_start, sentence_end, _CLAUSE_BOUNDARY_RE
        ):
            clause = masked_text[clause_start:clause_end]
            if count_cjk_relevant(clause) >= min_cjk:
                spans.add((clause_start, clause_end))

    return [Segment(start, end, masked_text[start:end]) for start, end in sorted(spans)]


def line_column(text: str, offset: int) -> tuple[int, int]:
    line = text.count("\n", 0, offset) + 1
    previous_newline = text.rfind("\n", 0, offset)
    column = offset + 1 if previous_newline < 0 else offset - previous_newline
    return line, column

File: internal/text/text.go
// Package text contains the Markdown masking and CJK segmentation shared by
// the quality-gate scanners.
package text

import (
	"sort"
	"unicode"
)

// Segment is a candidate sentence or clause, with rune offsets into the input.
type Segment struct {
	Start int
	End   int
	Text  string
}

// MaskMarkdownNonProse replaces fenced code, inline code, and URLs with spaces
// while retaining every rune position and newline. When includeCode is true,
// only URLs are masked.
func MaskMarkdownNonProse(input string, includeCode bool) string {
	runes := []rune(input)
	if !includeCode {
		maskFencedCode(runes)
		maskInlineCode(runes)
	}
	maskURLs(runes)
	return string(runes)
}

func maskFencedCode(runes []rune) {
	var fenceChar rune
	fenceLen := 0

	for lineStart := 0; lineStart < len(runes); {
		lineEnd := lineStart
		for lineEnd < len(runes) && !isSplitLineBoundary(runes[lineEnd]) {
			lineEnd++
		}
		nextLine := lineEnd
		if nextLine < len(runes) {
			nextLine++
			if runes[lineEnd] == '\r' && nextLine < len(runes) && runes[nextLine] == '\n' {
				nextLine++
			}
		}

		first := lineStart
		for first < lineEnd && isPythonSpace(runes[first]) {
			first++
		}
		runChar, runLen := fenceAt(runes, first, lineEnd)

		if fenceChar == 0 {
			if runLen >= 3 {
				fenceChar = runChar
				fenceLen = runLen
				blankExceptNewline(runes, lineStart, nextLine)
			}
		} else {
			blankExceptNewline(runes, lineStart, nextLine)
			if runChar == fenceChar && runLen >= fenceLen {
				fenceChar = 0
				fenceLen = 0
			}
		}

		lineStart = nextLine
	}
}

func fenceAt(runes []rune, start, end int) (rune, int) {
	if start >= end || (runes[start] != '`' && runes[start] != '~') {
		return 0, 0
	}
	char := runes[start]
	index := start + 1
	for index < end && runes[index] == char {
		index++
	}
	return char, index - start
}

func blankExceptNewline(runes []rune, start, end int) {
	for index := start; index < end; index++ {
		if runes[index] != '\n' {
			runes[index] = ' '
		}
	}
}

func isSplitLineBoundary(r rune) bool {
	return r == '\n' || r == '\r' || r == '\v' || r == '\f' ||
		(r >= 0x1c && r <= 0x1e) || r == 0x85 || r == 0x2028 || r == 0x2029
}

// maskInlineCode mirrors Python's (`+)([^\n]*?)\1 without relying on RE2
// backreferences. Python's opening quantifier is greedy but may backtrack; the
// descending run lengths below preserve that small edge-case of the reference
// implementation as well as the usual Markdown cases.
func maskInlineCode(runes []rune) {
	for index := 0; index < len(runes); {
		if runes[index] != '`' {
			index++
			continue
		}

		openingLen := backtickRun(runes, index)
		matchEnd := inlineCodeEnd(runes, index, openingLen)
		if matchEnd < 0 {
			index++
			continue
		}
		blankExceptNewline(runes, index, matchEnd)
		index = matchEnd
	}
}

func backtickRun(runes []rune, start int) int {
	end := start
	for end < len(runes) && runes[end] == '`' {
		end++
	}
	return end - start
}

func inlineCodeEnd(runes []rune, start, openingLen int) int {
	for runLen := openingLen; runLen >= 1; runLen-- {
		contentStart := start + runLen
		for index := contentStart; index < len(runes); index++ {
			if runes[index] == '\n' {
				break
			}
			if hasBacktickRun(runes, index, runLen) {
				return index + runLen
			}
		}
	}
	return -1
}

func hasBacktickRun(runes []rune, start, length int) bool {
	if start+length > len(runes) {
		return false
	}
	for index := start; index < start+length; index++ {
		if runes[index] != '`' {
			return false
		}
	}
	return true
}

func maskURLs(runes []rune) {
	for index := 0; index < len(runes); {
		prefixLen := urlPrefixLength(runes, index)
		if prefixLen == 0 {
			index++
			continue
		}
		end := index + prefixLen
		if end == len(runes) || isURLTerminator(runes[end]) {
			index++
			continue
		}
		for end < len(runes) && !isURLTerminator(runes[end]) {
			end++
		}
		blankExceptNewline(runes, index, end)
		index = end
	}
}

func urlPrefixLength(runes []rune, start int) int {
	if start+7 <= len(runes) && equalASCII(runes[start:start+7], "http://") {
		return 7
	}
	if start+8 <= len(runes) && equalASCII(runes[start:start+8], "https://") {
		return 8
	}
	return 0
}

func equalASCII(runes []rune, ascii string) bool {
	if len(runes) != len(ascii) {
		return false
	}
	for index, want := range ascii {
		if runes[index] != want {
			return false
		}
	}
	return true
}

func isURLTerminator(r rune) bool {
	return isPythonSpace(r) || r == '<' || r == '>' || r == '(' || r == ')'
}

// IsCJKRelevant reports whether r belongs to the broad candidate range used
// for segment detection. This is intentionally separate from CJClassifier's
// narrower scoring range.
func IsCJKRelevant(r rune) bool {
	return (r >= 0x3040 && r <= 0x30ff) ||
		(r >= 0x3400 && r <= 0x4dbf) ||
		(r >= 0x4e00 && r <= 0x9fff) ||
		(r >= 0xf900 && r <= 0xfaff) ||
		(r >= 0x20000 && r <= 0x323af)
}

// CountCJKRelevant counts runes in the broad candidate range.
func CountCJKRelevant(input string) int {
	count := 0
	for _, r := range input {
		if IsCJKRelevant(r) {
			count++
		}
	}
	return count
}

// IterCJSegments returns sentence and clause spans with at least minCJK
// relevant runes. Both granularities are retained so embedded Chinese clauses
// can be found inside otherwise Japanese sentences.
func IterCJSegments(input string, minCJK int) []Segment {
	runes := []rune(input)
	seen := make(map[[2]int]struct{})

	sentenceSpans := splitSpans(runes, 0, len(runes), isSentenceBoundary)
	for _, span := range sentenceSpans {
		if CountCJKRelevant(string(runes[span[0]:span[1]])) >= minCJK {
			seen[span] = struct{}{}
		}
		for _, clause := range splitSpans(runes, span[0], span[1], isClauseBoundary) {
			if CountCJKRelevant(string(runes[clause[0]:clause[1]])) >= minCJK {
				seen[clause] = struct{}{}
			}
		}
	}

	spans := make([][2]int, 0, len(seen))
	for span := range seen {
		spans = append(spans, span)
	}
	sort.Slice(spans, func(i, j int) bool {
		if spans[i][0] != spans[j][0] {
			return spans[i][0] < spans[j][0]
		}
		return spans[i][1] < spans[j][1]
	})

	segments := make([]Segment, 0, len(spans))
	for _, span := range spans {
		segments = append(segments, Segment{
			Start: span[0],
			End:   span[1],
			Text:  string(runes[span[0]:span[1]]),
		})
	}
	return segments
}

func splitSpans(runes []rune, start, end int, boundary func(rune) bool) [][2]int {
	spans := make([][2]int, 0)
	cursor := start
	for index := start; index < end; {
		if !boundary(runes[index]) {
			index++
			continue
		}
		boundaryEnd := index + 1
		for boundaryEnd < end && boundary(runes[boundaryEnd]) {
			boundaryEnd++
		}
		if trimmed, ok := trimSpan(runes, cursor, index); ok {
			spans = append(spans, trimmed)
		}
		cursor = boundaryEnd
		index = boundaryEnd
	}
	if trimmed, ok := trimSpan(runes, cursor, end); ok {
		spans = append(spans, trimmed)
	}
	return spans
}

func trimSpan(runes []rune, start, end int) ([2]int, bool) {
	for start < end && isPythonSpace(runes[start]) {
		start++
	}
	for end > start && isPythonSpace(runes[end-1]) {
		end--
	}
	return [2]int{start, end}, start < end
}

func isSentenceBoundary(r rune) bool {
	switch r {
	case '\n', '。', '！', '？', '!', '?':
		return true
	default:
		return false
	}
}

func isClauseBoundary(r rune) bool {
	switch r {
	case '、', '，', ',', ':', '：', '；', ';', '（', '）', '(', ')', '［', '］', '[', ']', '{', '}':
		return true
	default:
		return false
	}
}

// LineColumn converts a rune offset to the one-based line and column used by
// the Python implementation.
func LineColumn(input string, offset int) (line, column int) {
	runes := []rune(input)
	if offset < 0 {
		offset = 0
	}
	if offset > len(runes) {
		offset = len(runes)
	}
	line = 1
	previousNewline := -1
	for index := 0; index < offset; index++ {
		if runes[index] == '\n' {
			line++
			previousNewline = index
		}
	}
	if previousNewline < 0 {
		return line, offset + 1
	}
	return line, offset - previousNewline
}

func isPythonSpace(r rune) bool {
	// Python's str.isspace()/re \s include the ASCII information separators;
	// unicode.IsSpace covers the Unicode White_Space property.
	return unicode.IsSpace(r) || (r >= 0x1c && r <= 0x1f)
}
